### ISSUE TITLE
Middleware function gets recreated every function call

### ISSUE THREAD

**Issue by utrack** [pressly/chi#5]
Steps to reproduce:

Use this middleware func:
```
func CtxMiddleware(next chi.Handler) chi.Handler {
  fmt.Println("mw recreated!")
  return chi.HandlerFunc(func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
    ctx = context.WithValue(ctx, "key", "value")
    next.ServeHTTPC(ctx, w, r)
  })
}
```

"mw recreated!" should be echoed once on router.Use() call, but it echoes every time a request is handled.

**Comment by pkieltyka** [comment]
@utrack thanks for the report. I noticed the same thing the other day. I'll be pushing a fix shortly.

**Comment by GeertJohan** [comment]
On a related note, can we type `func(next chi.Handler) chi.Handler` ?

I use `New..` methods for some middleware to add stuff to them, which results in:
```go
func NewSomethingCtx(stuff *Stuff)  func(next chi.Handler) chi.Handler  {
	return func(next chi.Handler) chi.Handler {
		return chi.HandlerFunc(func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
			// actual middleware
		})
	}
}
```

If would be nice if we could let `NewSomethingCtx` return a `chi.Middleware`

**Comment by pkieltyka** [comment]
closed in fffe601cb0dece53671b9332886a6a0028d2dae4

**Comment by utrack** [comment]
Thank you! :+1:
